src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -18,13 +19,50 @@ export interface Comparison {
   expected: string; // The string representation of the 'expected' value (possibly with markup)
 }
 
+/** FloatTest is a function that compares whether the two floating point numbers match. */
+export interface FloatTest {
+  (got: number, expected: number): boolean;
+}
+
 /** Comparator is a function that compares whether the provided value matches an expectation. */
 export interface Comparator {
-  (got: Value): Comparison;
+  (got: Value, cmpFloats: FloatTest): Comparison;
+}
+
+/**
+ * @returns a FloatTest that returns true iff the two numbers are equal to, or
+ * less than the specified absolute error threshold.
+ */
+export function absThreshold(diff: number): FloatTest {
+  return (got, expected) => {
+    if (got === expected) {
+      return true;
+    }
+    if (!Number.isFinite(got) || !Number.isFinite(expected)) {
+      return false;
+    }
+    return Math.abs(got - expected) <= diff;
+  };
+}
+
+/**
+ * @returns a FloatTest that returns true iff the two numbers are within or
+ * equal to the specified ULP threshold value.
+ */
+export function ulpThreshold(ulp: number): FloatTest {
+  return (got, expected) => {
+    if (got === expected) {
+      return true;
+    }
+    if (!Number.isFinite(got) || !Number.isFinite(expected)) {
+      return false;
+    }
+    return diffULP(got, expected) <= ulp;
```

Comment:
I wonder if diffULP should be updated to return Infinity when either argument is NaN or Infinity.
Then callers could be simplified, like here.

Reply:
Done